Add device state varz

didovesei opened the pull request.

grafnu commented on forch/forchestrator.py:
```diff
@@ -208,9 +208,16 @@ def _initialize_orchestration(self):
+        if self._metrics:
+            update_device_state_varz = (lambda mac, state: self._metrics.update_var(
+                'device_state', state, labels=[mac]))
+            self._port_state_manager = PortStateManager(
```
Refactor this out so constructor call isn't duplicated

didovesei replied:
Done

grafnu commented on forch/port_state_manager.py:
```diff
@@ -22,30 +23,38 @@ def register(func):
 
     TRANSITIONS = {
-        AUTHENTICATED: {
+        UNAUTHENTICATED: {
```
This doesn't seem right. We should only be getting to this point after authentication, no?

didovesei replied:
So "authenticated" is more like a transient state. Conceptually device is authenticated, but then depending the static port behavior, the port will be set to either sequestered or operational once authenticated. So the port will never be in a state  called "authenticated"...

grafnu commented on forch/port_state_manager.py:
```diff
@@ -154,13 +184,10 @@ def _handle_authenticated_device(self, mac, device_behavior, static):
+                self._state_machines[mac].handle_port_behavior(port_behavior)
 
-    def _handle_unauthenticated_device(self, mac, static):
+    def _handle_deauthenticated_device(self, mac, static):
```
comment no longer matches function name

didovesei replied:
updated

grafnu commented on forch/port_state_manager.py:
```diff
@@ -188,10 +218,15 @@ def _handle_port_behavior(self, mac, port_behavior):
             state_machine.handle_port_behavior(port_behavior)
 
+    def _handle_unauthenticated_state(self, mac):
+        """Handle unauthenticated state"""
```
no need for comment on private function (and it's highly redundant)

didovesei replied:
removed

grafnu commented on forch/port_state_manager.py:
```diff
@@ -104,15 +105,22 @@ def _handle_infracted_state(self):
-        self._process_device_behavior = process_device_behavior
-        self._get_vlan_from_segment = get_vlan_from_segment
+        self._process_device_placement = (
+            lambda *args, **kwargs: process_device_placement(*args, **kwargs)
```
why are you doing it this way?  This seems very fragile -- any change to the arguments will then implicitly effect all these functions. This really doesn't feel right (not sure what it's doing, but...). You should explicitly use the arguments that need to be used unless you're doing something that's very "generic" -- typically any use of *args or **kwargs is a bad idea...

didovesei replied:
I think previsouly the number of args passed to the lambda function may differ in different cases. I have udpated it so that it does use args and kwargs. PTAL...

grafnu commented on forch/port_state_manager.py:
```diff
@@ -114,10 +114,10 @@ def __init__(self, process_device_placement, process_device_behavior, get_vlan_f
         self._dynamic_device_behaviors = {}
         self._process_device_placement = (
-            lambda *args, **kwargs: process_device_placement(*args, **kwargs)
+            lambda mac, placement, static: process_device_placement(mac, placement, static)
```
why do you need the lambda function at all?  Isn't this equivalent to just saying:

`self._process_device_placement = process_device_placement`

??  Maybe I'm missing what this is doing.

didovesei replied:
So that's mainly because the functions passed in (e.g., process_device_placement) may be None. So using a lambda and if/else here is equivalent to:
```
def _process_device_placement(mac, placement, static):
    if process_device_placement:
        process_device_placement(mac, placement, static)
```

grafnu commented:
I don't understand how that works... What am I missing with my simple test
below?  Also, if it's None then you can still do the assignment and just
deal with it where the function is called? (Not saying that's better
necessarily, just trying to clarify my understanding.)

~$ python3
Python 3.8.5 (default, Aug  2 2020, 15:09:07)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> bar = lambda x : foo(x)
>>> bar(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <lambda>
NameError: name 'foo' is not defined
>>> foo = print
>>> bar(2)
2

On Tue, Nov 24, 2020 at 5:13 PM Yufeng Duan <notifications@github.com>
wrote:

> *@didovesei* commented on this pull request.
> ------------------------------
>
> In forch/port_state_manager.py
> <https://github.com/faucetsdn/forch/pull/220#discussion_r530044674>:
>
> > @@ -114,10 +114,10 @@ def __init__(self, process_device_placement, process_device_behavior, get_vlan_f
>          self._static_device_behaviors = {}
>          self._dynamic_device_behaviors = {}
>          self._process_device_placement = (
> -            lambda *args, **kwargs: process_device_placement(*args, **kwargs)
> +            lambda mac, placement, static: process_device_placement(mac, placement, static)
>
> So that's mainly because the functions passed in (e.g.,
> process_device_placement) may be None. So using a lambda and if/else here
> is equivalent to:
>
> def _process_device_placement(mac, placement, static):
>     if process_device_placement:
>         process_device_placement(mac, placement, static)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/faucetsdn/forch/pull/220#discussion_r530044674>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPDYAYJRU26Y4PX636CLSRRK35ANCNFSM4UATXRTA>
> .
>

didovesei commented:
>if it's None then you can still do the assignment and just deal with it where the function is called? (Not saying that's better necessarily, just trying to clarify my understanding.) 

Yes, that's right... We can check if it is None while calling it. So my idea was that instead of checking if it is None every time it is called, we can build that check in the function itself.

So let me try to explain with the two examples below...

First example:
```
>>> foo=None
>>> bar=foo
>>> bar(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object is not callable
>>> 
```

If we check if foo is None or in the function then we do not need to check if bar is None every time bar is called:
```
>>> bar = lambda x: foo(x) if foo else None
>>> bar(2)
>>> 
```

Maybe this is not intuitive as the code shows... Should I replace the lambda implementation with a None check when the functions are called?

grafnu commented:
I think so -- the reason is that when somebody is trying to understand
what's going on it's usually at the point of when the function is called...
and that's where you'd want to be able to easily see "is this function
defined or not" -- if you abstract it away so that it's always a defined
function it becomes much more indirect how that happens. It should be more
or less the same amount of code, just where the if statement is.  If the
check would be handled in several places that might change things (DRY).

Ah, the part that confused me was the "if" was on the following line --
that's python to blame I think -- that's a pretty bad language construct
IMHO, but that's a separate rant.


On Tue, Nov 24, 2020 at 5:36 PM Yufeng Duan <notifications@github.com>
wrote:

> if it's None then you can still do the assignment and just deal with it
> where the function is called? (Not saying that's better necessarily, just
> trying to clarify my understanding.)
>
> Yes, that's right... We can check if it is None while calling it. So my
> idea was that instead of checking if it is None every time it is called, we
> can build that check in the function itself.
>
> So let me try to explain with the two examples below...
>
> First example:
>
> >>> foo=None
> >>> bar=foo
> >>> bar(2)
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> TypeError: 'NoneType' object is not callable
> >>>
>
> If we check if foo is None or in the function then we do not need to check
> if bar is None every time bar is called:
>
> >>> bar = lambda x: foo(x) if foo else None
> >>> bar(2)
> >>>
>
> Maybe this is not intuitive as the code shows... Should I replace the
> lambda implementation with a None check when the functions are called?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/faucetsdn/forch/pull/220#issuecomment-733406578>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPDYQONAFNGANCKL422LSRRNQDANCNFSM4UATXRTA>
> .
>

didovesei commented:
> I think so -- the reason is that when somebody is trying to understand what's going on it's usually at the point of when the function is called... and that's where you'd want to be able to easily see "is this function defined or not" -- if you abstract it away so that it's always a defined function it becomes much more indirect how that happens. It should be more or less the same amount of code, just where the if statement is. If the check would be handled in several places that might change things (DRY). Ah, the part that confused me was the "if" was on the following line -- that's python to blame I think -- that's a pretty bad language construct IMHO, but that's a separate rant.
> […](#)
> On Tue, Nov 24, 2020 at 5:36 PM Yufeng Duan ***@***.***> wrote: if it's None then you can still do the assignment and just deal with it where the function is called? (Not saying that's better necessarily, just trying to clarify my understanding.) Yes, that's right... We can check if it is None while calling it. So my idea was that instead of checking if it is None every time it is called, we can build that check in the function itself. So let me try to explain with the two examples below... First example: >>> foo=None >>> bar=foo >>> bar(2) Traceback (most recent call last): File "<stdin>", line 1, in <module> TypeError: 'NoneType' object is not callable >>> If we check if foo is None or in the function then we do not need to check if bar is None every time bar is called: >>> bar = lambda x: foo(x) if foo else None >>> bar(2) >>> Maybe this is not intuitive as the code shows... Should I replace the lambda implementation with a None check when the functions are called? — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#220 (comment)](https://github.com/faucetsdn/forch/pull/220#issuecomment-733406578)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAIEPDYQONAFNGANCKL422LSRRNQDANCNFSM4UATXRTA> .

Got it. Removed the usage of lambda function and updated the other pieces in the same class accordingly. PTAL..
